src/services/Store/store/account.slice.ts

```diff
@@ -216,6 +242,96 @@ export function* handleAddAccounts({ payload }: PayloadAction<IAccount[]>) {
   }
 }
 
+export function* addNewAccountsWorker({
+  payload: { networkId, accountType, newAccounts }
+}: PayloadAction<{
+  networkId: NetworkId;
+  accountType?: WalletId;
+  newAccounts: IAccountAdditionData[];
+}>) {
+  const accounts: StoreAccount[] = yield select(getStoreAccounts);
+  const networks: Network[] = yield select(selectNetworks);
+  const assets: Asset[] = yield select(getAssets);
+  const contacts = yield select(selectContacts);
+  const contracts = yield select(selectContracts);
+
+  const network = getNetworkById(networkId, networks);
+  if (!network || newAccounts.length === 0) return;
+  const accountsToAdd = newAccounts.filter(
+    ({ address }) => !getAccountByAddressAndNetworkName(accounts)(address, networkId)
+  );
+  const walletType = accountType! === WalletId.WEB3 ? getWeb3Config().id : accountType!;
+  const newAsset = getNewDefaultAssetTemplateByNetwork(assets)(network);
+  const newRawAccounts = accountsToAdd.map(({ address, dPath }) => ({
+    uuid: generateDeterministicAddressUUID(networkId, address),
+    address,
+    networkId,
+    wallet: walletType,
+    dPath,
+    assets: [{ uuid: newAsset.uuid, balance: '0', mtime: Date.now() }],
+    transactions: [],
+    mtime: 0,
+    favorite: false,
+    isPrivate: undefined
+  }));
+  if (newRawAccounts.length === 0) {
+    yield put(displayNotification({ templateName: NotificationTemplates.walletsNotAdded }));
+    return;
+  }
+  const newLabels = findMultipleNextUnusedDefaultLabels(
+    newRawAccounts[0].wallet,
+    newRawAccounts.length
+  )(contacts);
+  const newContacts = newRawAccounts
+    .map((rawAccount, idx) => {
+      const existingContact = getContactByAddressAndNetworkId(contacts, contracts)(
+        rawAccount.address,
+        networkId
+      );
+      if (existingContact && existingContact.label === translateRaw('NO_LABEL')) {
+        return {
+          ...existingContact,
+          label: newLabels[idx]
+        };
+      } else if (!existingContact) {
+        return {
+          label: newLabels[idx],
+          address: rawAccount.address,
+          notes: '',
+          network: rawAccount.networkId,
+          uuid: generateUUID()
+        };
+      }
+      return undefined;
+    })
+    .filter((a) => a !== undefined) as ExtendedContact[];
+  if (newContacts.length > 0) {
+    yield put(createOrUpdateContacts(newContacts));
+  }
+  yield put(addAccounts(newRawAccounts));
+  yield put(scanTokens({ accounts: newRawAccounts }));
+  yield put(fetchMemberships(newRawAccounts));
```

Comment:
this could be done inside some `takeLatest` nah ?

Reply:
How? The format of the payload is not necessarily the same?

Reply:
Yeah I just saw that scanToken don't have the same payload